src/visitable.cpp

```diff
@@ -72,21 +72,30 @@ bool visitable<T>::has_item_with( const std::function<bool( const item & )> &fil
     } ) == VisitResponse::ABORT;
 }
 
+// Sums the two terms, being careful to not trigger any overflows.
+// Assumes that all terms are positive and base <= limit.
+static long accumulate_with_limit( long base, long addend, long limit ) {
+    if( addend > limit - base ) {
+        return limit;
+    }
+    return base + addend;
+}
+
 template <typename T>
 static int has_quality_internal( const T& self, const quality_id &qual, int level, int limit )
 {
     int qty = 0;
 
     self.visit_items( [&qual, level, &limit, &qty]( const item *e ) {
         if( e->get_quality( qual ) >= level ) {
-            qty += e->count_by_charges() ? e->charges : 1;
+            qty = accumulate_with_limit( qty, e->count_by_charges() ? e->charges : 1, limit );
```

Comment:
Significant: Brittle. Such overflow is a common fault in our codebase and this solution will become incomplete rapidly.

A more correct solution could be to define a type for charges and then implement arithmetic operators for it which prevent overflow via clamping. This would be helpful as it's then impossible to reintroduce the bug and we also get type safety.

We already have examples of such in the code base (volume) but something less complex would suffice here